networkx/algorithms/cluster.py

```diff
@@ -42,22 +46,47 @@ def triangles(G, nodes=None):
     6
     >>> print(nx.triangles(G))
     {0: 6, 1: 6, 2: 6, 3: 6, 4: 6}
-    >>> print(list(nx.triangles(G, (0, 1)).values()))
+    >>> print(list(nx.triangles(G, [0, 1]).values()))
     [6, 6]
 
     Notes
     -----
-    When computing triangles for the entire graph each triangle is counted
-    three times, once at each node.  Self loops are ignored.
+    Self loops are ignored.
 
     """
-    # If `nodes` represents a single node in the graph, return only its number
-    # of triangles.
-    if nodes in G:
-        return next(_triangles_and_degree_iter(G, nodes))[2] // 2
-    # Otherwise, `nodes` represents an iterable of nodes, so return a
-    # dictionary mapping node to number of triangles.
-    return {v: t // 2 for v, d, t, _ in _triangles_and_degree_iter(G, nodes)}
+    if nodes is not None:
+        # If `nodes` represents a single node, return only its number of triangles
+        if nodes in G:
+            return next(_triangles_and_degree_iter(G, nodes))[2] // 2
+
+        # if `nodes` is a container of nodes, then return a
+        # dictionary mapping node to number of triangles.
+        return {v: t // 2 for v, d, t, _ in _triangles_and_degree_iter(G, nodes)}
+
+    # if nodes is None, then compute triangles for the complete graph
+
+    # dict used to avoid visiting the same nodes twice
+    # this allows calculating/counting each triangle only once
+    later_neighbors = {}
+
+    # iterate over the nodes in a graph
+    for node, neighbors in G.adjacency():
+        later_neighbors[node] = {
+            n for n in neighbors if n not in later_neighbors and n is not node
+        }
+
+    # instantiate with a zero count for each node
+    # add 1 to the count if a nodes neighbor's neighbor is also a neighbor
+    triangle_counts = Counter(dict.fromkeys(G, 0))
```

Comment:
Why is instantiating an empty counter necessary? we are guaranteed to have all nodes in `later_neighbours` so an empty `Counter` should be ok I would have thought.

```suggestion
    Counter()
```

If for whatever reason the initialized Counter is necessary, you can avoid instantiating 2 dictionaries by doing
```python
defaultdict(int).fromkeys(G, 0)
```
instead, without much of a performance difference.

Reply:
It's not much, but it helps with ensuring that nodes without neighbors have zero counts. If we have `triangle_counts` initiated using blank `Counter`, there would have to be a separate loop to fill in zeros for nodes not in `triangle_counts`.

Reply:
All nodes appear in `G.adjacency()`, so all nodes appear in `later_neighbors` - if they have no neighbors they just appear with the empty set. So, you can avoid initializing the Counter by adding a statement inside the first block, but outside the second:

```python
triangle_counts = Counter(dict.fromkeys(G, 0))
for node1, neighbors in later_neighbors.items():
    if not neighbors:
        triangle_counts[node1] = 0
    for node2 in neighbors:
        ...
```
checking the boolean is optional, I guess.

Reply:
The performance change is interesting, but a more robust benchmarking would be needed for maintainers. This is one of the missing pieces in the current PR for it to be considered for merging. 

Also, your suggestion might have a better chance as a standalone PR because in this one I tried to combine the triangle count change with a more explicit handling of `nodes`/`nbunch` (the unresolved conversation above). I reverted that behavior to be consistent with the original code, but I guess it doesn't look very clean. This is not to say that your suggestions are not welcome for this PR :), but to say that after performance benchmarking there might still be an issue with the code readability.

Reply:
Looking closer at your suggestion though, I am not actually sure about the source of the performance improvement. 

```diff
- triangle_counts = Counter(dict.fromkeys(G, 0))
+ triangle_counts = Counter()
  for node1, neighbors in later_neighbors.items():
+   acc = 0
    for node2 in neighbors:
       third_nodes = neighbors & later_neighbors[node2]
       m = len(third_nodes)
+      acc += m
-      triangle_counts[node1] += m
       triangle_counts[node2] += m
       triangle_counts.update(third_nodes)
+   triangle_counts[node1] += acc
```

As you note, we are modifying the `node1` entry of `triangle_counts` less frequently, but this seems a bit like a micro-optimisation (see a talk by James Powell that has nice examples: https://youtu.be/Ix04KpZiUA8?t=754). Specifically, this *might* have performance boost for nodes that have a lot of neighbors if updating counter element is more expensive than updating a variable. However, in terms of readability (and this is subjective) the reader has to keep track of three lines (involving `acc`) to understand what's going on, which decreases readability (but again, this is subjective).

Reply:
Will leave it up to you, but my preference here would be the minor optimization for the very minor increase in code length. My thinking is that triangle counting will be a function used often, so speed will improve the user experience.
> might have performance boost for nodes that have a lot of neighbors

I think we can say more than that, variable updates are faster than dict lookups, and initializing a dictionary vs not is clearly slower. The worst case scenario for this algorithm is highly connected graphs - so improving that case will be beneficial since graphs that are not highly connected will be fast anyway.

But, with all that considered, it's still a minor thing :) If you are happier with your code as it is, it should be merged that way - I just wanted to mention this as it jumped out at me when I first read it.